libr/bin/format/mach0/coresymbolication.c

```diff
@@ -0,0 +1,306 @@
+		}
+		cursor += 2;
+	}
+	free (fileline);
+}
+
+RCoreSymCacheElement *r_coresym_cache_element_new(RBinFile *bf, RBuffer *buf, ut64 off, int bits) {
+	RCoreSymCacheElement *result = NULL;
+	ut8 *b = NULL;
+	RCoreSymCacheElementHdr *hdr = r_coresym_cache_element_header_new (buf, off, bits);
+	if (!hdr) {
+		return NULL;
+	}
+	if (hdr->version != 1) {
+		eprintf ("Unsupported CoreSymbolication cache version (%d)\n", hdr->version);
+		goto beach;
+	}
+	if (hdr->size == 0 || hdr->size > r_buf_size (buf) - off) {
+		eprintf ("Corrupted CoreSymbolication header: size out of bounds (0x%x)\n", hdr->size);
+		goto beach;
+	}
+	result = R_NEW0 (RCoreSymCacheElement);
+	if (!result) {
+		goto beach;
+	}
+	result->hdr = hdr;
+	b = malloc (hdr->size);
+	if (!b) {
+		goto beach;
+	}
+	if (r_buf_read_at (buf, off, b, hdr->size) != hdr->size) {
+		goto beach;
+	}
+	ut8 *end = b + hdr->size;
+	if (hdr->file_name_off) {
+		const char *file_name_str = (const char *)b + hdr->file_name_off;
+		result->file_name = r_str_ndup (file_name_str,
+				r_str_nlen (file_name_str, 1024));
+	}
+	if (hdr->version_off) {
+		const char *binary_version_str = (const char *)b + hdr->version_off;
+		result->binary_version = r_str_ndup (binary_version_str,
+				r_str_nlen (binary_version_str, 16));
+	}
+	const size_t word_size = bits / 8;
+	ut64 page_zero_size = 0;
+	size_t page_zero_idx = 0;
+	if (hdr->n_segments > 0) {
+		result->segments = R_NEWS0 (RCoreSymCacheElementSegment, hdr->n_segments);
+		if (!result->segments) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + R_CS_EL_OFF_SEGS;
+		for (i = 0; i < hdr->n_segments && cursor < end; i++) {
+			RCoreSymCacheElementSegment *seg = &result->segments[i];
+			seg->paddr = seg->vaddr = r_read_le64 (cursor);
+			cursor += 8;
+			if (cursor >= end) {
+				break;
+			}
+			seg->size = seg->vsize = r_read_le64 (cursor);
+			cursor += 8;
+			if (cursor >= end) {
+				break;
+			}
+			r_str_ncpy (seg->name, (const char *)cursor, 16);
+			cursor += 16;
+
+			if (!strcmp (seg->name, "__PAGEZERO")) {
+				page_zero_size = seg->size;
+				page_zero_idx = i;
+				seg->paddr = seg->vaddr = 0;
+				seg->size = 0;
+			}
+		}
+		for (i = 0; i < hdr->n_segments && page_zero_size > 0; i++) {
+			if (i == page_zero_idx) {
+				continue;
+			}
+			RCoreSymCacheElementSegment *seg = &result->segments[i];
+			if (seg->vaddr < page_zero_size) {
+				seg->vaddr += page_zero_size;
+			}
+		}
+	}
+	ut64 start_of_sections = (ut64)hdr->n_segments * R_CS_EL_SIZE_SEG + R_CS_EL_OFF_SEGS;
+	if (hdr->n_sections > 0) {
+		result->sections = R_NEWS0 (RCoreSymCacheElementSection, hdr->n_sections);
+		if (!result->sections) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_sections;
+		for (i = 0; i < hdr->n_sections && cursor < end; i++) {
+			ut8 *sect_start = cursor;
+			RCoreSymCacheElementSection *sect = &result->sections[i];
+			sect->vaddr = sect->paddr = r_read_ble (cursor, false, bits);
+			if (sect->vaddr < page_zero_size) {
```

Comment:
sounds like a hack, but if it works im fine with it for now :D

Reply:
yeah it's a hack. there's no info to do that better than this. there's only one address which apparently is an offset for the useful segments (the ones which contain symbols), but it's a virtual address for the other segments.